File: backend/app/api/routes/positions.py
"""Positions routes — open a new wheel and list all positions."""

import datetime
import uuid
from decimal import Decimal

import structlog
from fastapi import APIRouter, Depends
from fastapi.responses import JSONResponse
from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy.orm import selectinload

from app.api.schemas import (
    CreatePositionRequest,
    CreatePositionResponse,
    FieldError,
    PositionListItemResponse,
    ValidationErrorResponse,
)
from app.core.costbasis import CspLegInput, calculate_initial_csp_basis
from app.core.lifecycle import OpenWheelInput, ValidationError, open_wheel
from app.core.types import LegAction, LegRole, OptionType, StrategyType, WheelStatus
from app.db import get_session
from app.models import CostBasisSnapshot, Leg, Position

logger = structlog.get_logger(__name__)

router = APIRouter()


@router.get("/positions", response_model=list[PositionListItemResponse])
async def list_positions(
    session: AsyncSession = Depends(get_session),
) -> list[PositionListItemResponse]:
    logger.debug("list_positions_query_start")

    stmt = select(Position).options(
        selectinload(Position.legs),
        selectinload(Position.cost_basis_snapshots),
    )
    result = await session.execute(stmt)
    positions = result.scalars().all()

    logger.debug("list_positions_query_complete", count=len(positions))

    items: list[PositionListItemResponse] = []
    for position in positions:
        active_leg = _active_leg(position)
        latest_snapshot = _latest_snapshot(position)

        strike = active_leg.strike if active_leg else None
        expiration = active_leg.expiration if active_leg else None
        dte = (expiration - datetime.date.today()).days if expiration else None

        items.append(
            PositionListItemResponse(
                id=position.id,
                ticker=position.ticker,
                phase=position.phase,
                status=position.status,
                strike=strike,
                expiration=expiration,
                dte=dte,
                premium_collected=(
                    latest_snapshot.total_premium_collected if latest_snapshot else Decimal(0)
                ),
                effective_cost_basis=(
                    latest_snapshot.basis_per_share if latest_snapshot else Decimal(0)
                ),
            )
        )

    # Positions with a known DTE sort ascending; null-DTE positions sort last.
    items.sort(key=_dte_sort_key)

    logger.info("positions_listed", count=len(items))
    return items


def _dte_sort_key(item: PositionListItemResponse) -> tuple[bool, int]:
    """Sort key: positions with DTE ascending first, null-DTE last."""
    return (item.dte is None, item.dte if item.dte is not None else 0)


def _active_leg(position: Position) -> Leg | None:
    open_legs = [leg for leg in position.legs if leg.action == LegAction.open]
    if not open_legs:
        return None
    return max(open_legs, key=lambda leg: leg.fill_date)


def _latest_snapshot(position: Position) -> CostBasisSnapshot | None:
    if not position.cost_basis_snapshots:
        return None
    return max(position.cost_basis_snapshots, key=lambda s: s.snapshot_at)


@router.post(
    "/positions",
    response_model=CreatePositionResponse,
    status_code=201,
    responses={400: {"model": ValidationErrorResponse}},
)
async def create_position(
    body: CreatePositionRequest,
    session: AsyncSession = Depends(get_session),
) -> CreatePositionResponse | JSONResponse:
    fill_date = body.fill_date or datetime.date.today()
    now = datetime.datetime.now(datetime.UTC).replace(tzinfo=None)

    logger.debug(
        "create_position_inputs",
        ticker=body.ticker,
        strike=str(body.strike),
        expiration=str(body.expiration),
        contracts=body.contracts,
        premium_per_contract=str(body.premium_per_contract),
        fill_date=str(fill_date),
    )

    try:
        lifecycle_result = open_wheel(
            OpenWheelInput(
                ticker=body.ticker,
                strike=body.strike,
                expiration=body.expiration,
                contracts=body.contracts,
                premium_per_contract=body.premium_per_contract,
                fill_date=fill_date,
                reference_date=datetime.date.today(),
            )
        )
    except ValidationError as exc:
        logger.info(
            "position_validation_failed",
            field=exc.field,
            code=exc.code,
            ticker=body.ticker,
        )
        return JSONResponse(
            status_code=400,
            content=ValidationErrorResponse(
                detail=[FieldError(field=exc.field, code=exc.code, message=exc.message)]
            ).model_dump(),
        )

    logger.debug("lifecycle_validated", phase=lifecycle_result.phase.value)

    basis_result = calculate_initial_csp_basis(
        CspLegInput(
            strike=body.strike,
            premium_per_contract=body.premium_per_contract,
            contracts=body.contracts,
        )
    )
    logger.debug(
        "cost_basis_calculated",
        basis_per_share=str(basis_result.basis_per_share),
        total_premium_collected=str(basis_result.total_premium_collected),
    )

    position_id = uuid.uuid4()
    leg_id = uuid.uuid4()
    snapshot_id = uuid.uuid4()

    logger.debug(
        "db_write_start",
        position_id=str(position_id),
        leg_id=str(leg_id),
        snapshot_id=str(snapshot_id),
    )
    async with session.begin():
        position = Position(
            id=position_id,
            ticker=body.ticker,
            strategy_type=StrategyType.WHEEL,
            status=WheelStatus.active,
            phase=lifecycle_result.phase,
            opened_date=fill_date,
            account_id=body.account_id,
            thesis=body.thesis,
            notes=body.notes,
            tags=[],
            created_at=now,
            updated_at=now,
        )
        leg = Leg(
            id=leg_id,
            position_id=position_id,
            leg_role=LegRole.csp,
            action=LegAction.open,
            option_type=OptionType.put,
            strike=body.strike,
            expiration=body.expiration,
            contracts=body.contracts,
            premium_per_contract=body.premium_per_contract,
            fill_date=fill_date,
            created_at=now,
            updated_at=now,
        )
        snapshot = CostBasisSnapshot(
            id=snapshot_id,
            position_id=position_id,
            basis_per_share=basis_result.basis_per_share,
            total_premium_collected=basis_result.total_premium_collected,
            snapshot_at=now,
            created_at=now,
        )
        session.add_all([position, leg, snapshot])
        await session.flush()
        logger.debug("db_flush_ok", position_id=str(position_id))
        await session.refresh(position)
        await session.refresh(leg)
        await session.refresh(snapshot)

    logger.info(
        "position_created",
        position_id=str(position_id),
        ticker=body.ticker,
        phase=lifecycle_result.phase.value,
        contracts=body.contracts,
        strike=str(body.strike),
        basis_per_share=str(basis_result.basis_per_share),
        total_premium_collected=str(basis_result.total_premium_collected),
    )

    return CreatePositionResponse(
        position=position,
        leg=leg,
        cost_basis_snapshot=snapshot,
    )

File: src/main/services/positions.ts
// Service layer — DB access + core logic composition.
// No Electron or broker imports here.

import Database from 'better-sqlite3'
import Decimal from 'decimal.js'
import { randomUUID } from 'node:crypto'
import { calculateInitialCspBasis } from '../core/costbasis'
import { openWheel } from '../core/lifecycle'
import { localToday, makeSnapshotAt } from '../dates'
import { logger } from '../logger'
import type { CreatePositionPayload, CreatePositionResult } from '../schemas'

export { listPositions } from './list-positions'
export { getPosition } from './get-position'
export { closeCspPosition } from './close-csp-position'
export { expireCspPosition } from './expire-csp-position'
export { assignCspPosition } from './assign-csp-position'
export { openCoveredCallPosition } from './open-covered-call-position'
export { closeCoveredCallPosition } from './close-covered-call-position'
export { expireCcPosition } from './expire-cc-position'

// ---------------------------------------------------------------------------
// createPosition
// ---------------------------------------------------------------------------

export function createPosition(
  db: Database.Database,
  payload: CreatePositionPayload
): CreatePositionResult {
  const today = localToday()
  const fillDate = payload.fillDate ?? today
  const now = new Date().toISOString()

  logger.debug(
    { ticker: payload.ticker, strike: payload.strike, expiration: payload.expiration, fillDate },
    'create_position_inputs'
  )

  const strikeStr = String(payload.strike)
  const premiumStr = String(payload.premiumPerContract)

  const lifecycleResult = openWheel({
    ticker: payload.ticker,
    strike: strikeStr,
    expiration: payload.expiration,
    contracts: payload.contracts,
    premiumPerContract: premiumStr,
    fillDate,
    referenceDate: today
  })

  logger.debug({ phase: lifecycleResult.phase }, 'lifecycle_validated')

  const basisResult = calculateInitialCspBasis({
    strike: strikeStr,
    premiumPerContract: premiumStr,
    contracts: payload.contracts
  })

  logger.debug(basisResult, 'cost_basis_calculated')

  const positionId = randomUUID()
  const legId = randomUUID()
  const snapshotId = randomUUID()
  const strikeFormatted = new Decimal(strikeStr).toFixed(4)
  const premiumFormatted = new Decimal(premiumStr).toFixed(4)
  const basisFormatted = new Decimal(basisResult.basisPerShare).toFixed(4)
  const totalPremiumFormatted = new Decimal(basisResult.totalPremiumCollected).toFixed(4)

  db.transaction(() => {
    db.prepare(
      `INSERT INTO positions
        (id, ticker, strategy_type, status, phase, opened_date, account_id, notes, thesis, tags, created_at, updated_at)
       VALUES (?, ?, 'WHEEL', 'ACTIVE', ?, ?, ?, ?, ?, '[]', ?, ?)`
    ).run(
      positionId,
      payload.ticker,
      lifecycleResult.phase,
      fillDate,
      payload.accountId ?? null,
      payload.notes ?? null,
      payload.thesis ?? null,
      now,
      now
    )

    db.prepare(
      `INSERT INTO legs
        (id, position_id, leg_role, action, instrument_type, strike, expiration, contracts, premium_per_contract, fill_date, created_at, updated_at)
       VALUES (?, ?, 'CSP_OPEN', 'SELL', 'PUT', ?, ?, ?, ?, ?, ?, ?)`
    ).run(
      legId,
      positionId,
      strikeFormatted,
      payload.expiration,
      payload.contracts,
      premiumFormatted,
      fillDate,
      now,
      now
    )

    db.prepare(
      `INSERT INTO cost_basis_snapshots
        (id, position_id, basis_per_share, total_premium_collected, snapshot_at, created_at)
       VALUES (?, ?, ?, ?, ?, ?)`
    ).run(
      snapshotId,
      positionId,
      basisFormatted,
      totalPremiumFormatted,
      makeSnapshotAt(fillDate),
      now
    )
  })()

  logger.info(
    { positionId, ticker: payload.ticker, phase: lifecycleResult.phase, basisFormatted },
    'position_created'
  )

  return {
    position: {
      id: positionId,
      ticker: payload.ticker,
      phase: lifecycleResult.phase,
      status: 'ACTIVE',
      strategyType: 'WHEEL',
      openedDate: fillDate,
      closedDate: null,
      accountId: payload.accountId ?? null,
      notes: payload.notes ?? null,
      thesis: payload.thesis ?? null,
      tags: [],
      createdAt: now,
      updatedAt: now
    },
    leg: {
      id: legId,
      positionId,
      legRole: 'CSP_OPEN',
      action: 'SELL',
      instrumentType: 'PUT',
      strike: strikeFormatted,
      expiration: payload.expiration,
      contracts: payload.contracts,
      premiumPerContract: premiumFormatted,
      fillPrice: premiumFormatted,
      fillDate,
      createdAt: now,
      updatedAt: now
    },
    costBasisSnapshot: {
      id: snapshotId,
      positionId,
      basisPerShare: basisFormatted,
      totalPremiumCollected: totalPremiumFormatted,
      finalPnl: null,
      snapshotAt: now,
      createdAt: now
    }
  }
}
